File: pygame_client/stdb_client.py
import json
import os

import config
import http_backend
from http_backend import HttpError


class StdbError(Exception):
    """Raised when a SpacetimeDB HTTP call fails (network, auth, or server error)."""


def sql_escape(value: str) -> str:
    """Escapes a string literal for embedding into a raw SQL query."""
    return value.replace("'", "''")


_IDENTITY_STORAGE_KEY = "xianxia_stdb_identity"


class StdbClient:
    def __init__(self, server_url=None, database=None):
        self.server_url = (server_url or config.SERVER_URL).rstrip("/")
        self.database = database or config.DATABASE_NAME
        self.identity = None
        self.token = None

    def _load_cached_identity(self):
        if http_backend.IS_WEB:
            raw = http_backend.web_storage_get(_IDENTITY_STORAGE_KEY)
            return json.loads(raw) if raw else None
        if os.path.exists(config.IDENTITY_FILE):
            with open(config.IDENTITY_FILE, "r") as f:
                return json.load(f)
        return None

    def _save_cached_identity(self, data):
        if http_backend.IS_WEB:
            http_backend.web_storage_set(_IDENTITY_STORAGE_KEY, json.dumps(data))
        else:
            with open(config.IDENTITY_FILE, "w") as f:
                json.dump(data, f)

    async def get_or_create_identity(self):
        cached = self._load_cached_identity()
        if cached:
            self.identity = cached["identity"]
            self.token = cached["token"]
            return

        try:
            status, text = await http_backend.request("POST", f"{self.server_url}/v1/identity")
        except HttpError as exc:
            raise StdbError(f"Konnte keine Identity erzeugen: {exc}") from exc
        if status >= 300:
            raise StdbError(f"Identity-Erzeugung fehlgeschlagen (HTTP {status}): {text}")

        data = json.loads(text)
        self.identity = data["identity"]
        self.token = data["token"]
        self._save_cached_identity(data)

    def _headers(self):
        return {
            "Authorization": f"Bearer {self.token}",
            "Content-Type": "application/json",
        }

    async def sql(self, query):
        """Runs a SQL query, returns a list of dicts (rows of the first statement)."""
        try:
            status, text = await http_backend.request(
                "POST",
                f"{self.server_url}/v1/database/{self.database}/sql",
                headers=self._headers(),
                body=query,
            )
        except HttpError as exc:
            raise StdbError(f"SQL-Anfrage fehlgeschlagen: {exc}") from exc
        if status >= 300:
            raise StdbError(f"SQL-Anfrage fehlgeschlagen (HTTP {status}): {text}")

        statements = json.loads(text)
        if not statements:
            return []

        schema = statements[0]["schema"]
        raw_rows = statements[0]["rows"]
        columns = [el["name"]["some"] for el in schema["elements"]]

        rows = []
        for raw_row in raw_rows:
            rows.append(dict(zip(columns, raw_row)))
        return rows

    async def call_reducer(self, name, args):
        try:
            status, text = await http_backend.request(
                "POST",
                f"{self.server_url}/v1/database/{self.database}/call/{name}",
                headers=self._headers(),
                body=json.dumps(args),
            )
        except HttpError as exc:
            raise StdbError(f"Reducer '{name}' fehlgeschlagen: {exc}") from exc
        if status >= 300:
            raise StdbError(f"Reducer '{name}' fehlgeschlagen (HTTP {status}): {text}")

File: web_client/stdb_client.js
// Thin wrapper over SpacetimeDB's HTTP API. No WebSocket subscriptions -
// callers poll via sql() on their own schedule.

// sessionStorage (not localStorage): scoped per tab, not shared across tabs
// of the same origin. That means each new tab gets its own fresh identity
// automatically - the natural way to test two players at once with this
// client is to just open a second tab, rather than needing a second browser
// profile/incognito window. Reloading a tab keeps the same identity; closing
// it drops it (server-side: one Player row per identity, see Lib.cs).
const IDENTITY_STORAGE_KEY = "xianxia_stdb_identity";

class StdbError extends Error {}

function sqlEscape(value) {
  return value.replace(/'/g, "''");
}

class StdbClient {
  constructor(serverUrl, database) {
    this.serverUrl = serverUrl.replace(/\/+$/, "");
    this.database = database;
    this.identity = null;
    this.token = null;
  }

  _loadCachedIdentity() {
    const raw = sessionStorage.getItem(IDENTITY_STORAGE_KEY);
    return raw ? JSON.parse(raw) : null;
  }

  _saveCachedIdentity(data) {
    sessionStorage.setItem(IDENTITY_STORAGE_KEY, JSON.stringify(data));
  }

  async getOrCreateIdentity() {
    const cached = this._loadCachedIdentity();
    if (cached) {
      this.identity = cached.identity;
      this.token = cached.token;
      return;
    }

    let resp;
    try {
      resp = await fetch(`${this.serverUrl}/v1/identity`, { method: "POST" });
    } catch (exc) {
      throw new StdbError(`Konnte keine Identity erzeugen (Netzwerk/CORS): ${exc}`);
    }
    if (!resp.ok) {
      throw new StdbError(`Identity-Erzeugung fehlgeschlagen (HTTP ${resp.status}): ${await resp.text()}`);
    }

    const data = await resp.json();
    this.identity = data.identity;
    this.token = data.token;
    this._saveCachedIdentity(data);
  }

  _headers() {
    return {
      Authorization: `Bearer ${this.token}`,
      "Content-Type": "application/json",
    };
  }

  /** Runs a SQL query, returns an array of row objects (first statement only). */
  async sql(query) {
    let resp;
    try {
      resp = await fetch(`${this.serverUrl}/v1/database/${this.database}/sql`, {
        method: "POST",
        headers: this._headers(),
        body: query,
      });
    } catch (exc) {
      throw new StdbError(`SQL-Anfrage fehlgeschlagen (Netzwerk/CORS): ${exc}`);
    }
    if (!resp.ok) {
      throw new StdbError(`SQL-Anfrage fehlgeschlagen (HTTP ${resp.status}): ${await resp.text()}`);
    }

    const statements = await resp.json();
    if (!statements || statements.length === 0) return [];

    const { schema, rows: rawRows } = statements[0];
    const columns = schema.elements.map((el) => el.name.some);

    return rawRows.map((rawRow) => {
      const row = {};
      columns.forEach((col, i) => (row[col] = rawRow[i]));
      return row;
    });
  }

  async callReducer(name, args) {
    let resp;
    try {
      resp = await fetch(`${this.serverUrl}/v1/database/${this.database}/call/${name}`, {
        method: "POST",
        headers: this._headers(),
        body: JSON.stringify(args),
      });
    } catch (exc) {
      throw new StdbError(`Reducer '${name}' fehlgeschlagen (Netzwerk/CORS): ${exc}`);
    }
    if (!resp.ok) {
      throw new StdbError(`Reducer '${name}' fehlgeschlagen (HTTP ${resp.status}): ${await resp.text()}`);
    }
  }
}
